File: retrocpu_boot_monitor/mn1613/test/handshake/handshake_hex_keyboard_test.ts
/**
 * g_bios_hex_key_get（CPU→IO コマンド 11h）
 * 根拠: HandShake.mdc「16進キー入力取得」/ boot_monitor.mdc / test_framework.mdc
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

const BASE_REGS = {
  R2: 0x2222,
  R3: 0x3333,
  R4: 0x4444,
} as const;

/** ユーザ RAM 上の結果バッファ（ワードアドレス） */
const KEY_BUF = 0x1800;

/** フリーモード（ハンドシェイク 10h の値） */
const MODE_FREE = 1;

/** 列 0–7 の押下ビットマップ */
const HEX_COLS = [0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80] as const;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(mn1613MonHandshakeSettings, import.meta.url),
);

/**
 * g_main 済み＋ ioMock handshake で 1 ケースを実行する。
 * @param fn 本体（session / mock 利用可）
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * g_bios_hex_key_get を呼び、CPU→IO 1 トランザクションと並行する。
 * @param mock IO モック
 * @param bufWordAddr 結果バッファ先頭
 */
async function callHexKey(
  mock: IoBoardHandshakeMock,
  bufWordAddr: number,
): Promise<void> {
  await Promise.all([
    session.call("g_bios_hex_key_get", {
      registers: { ...BASE_REGS, R0: bufWordAddr },
    }),
    mock.handleOneRequest(),
  ]);
}

/**
 * バッファ 8 ワードの下位 8bit を読む。
 * @param s セッション
 * @param wordAddr 先頭
 * @returns 列 0–7
 */
function readKeyBuf(s: Mn1613AsmSession, wordAddr: number): number[] {
  return Array.from({ length: 8 }, (_, i) => s.readWord(wordAddr + i) & 0xff);
}

test("フリーモードで列 0–7 のビットマップがバッファに入る", async () => {
  await withCase(async (s, mock) => {
    mock.state.mode = MODE_FREE;
    mock.setHexKeys([...HEX_COLS]);
    await callHexKey(mock, KEY_BUF);
    s.expectRegisters({ R0: 0 });
    expect(readKeyBuf(s, KEY_BUF)).toEqual([...HEX_COLS]);
  });
});

test("モニターモードでは 01h を返し列は 0", async () => {
  await withCase(async (s, mock) => {
    mock.setHexKeys([...HEX_COLS]);
    await callHexKey(mock, KEY_BUF);
    s.expectRegisters({ R0: 1 });
    expect(readKeyBuf(s, KEY_BUF)).toEqual([0, 0, 0, 0, 0, 0, 0, 0]);
  });
});

test("R3/R4 は呼び出しの前後で保たれる", async () => {
  await withCase(async (s, mock) => {
    mock.state.mode = MODE_FREE;
    mock.setHexKeys([...HEX_COLS]);
    await callHexKey(mock, KEY_BUF);
    s.expectRegisters({ R0: 0, R3: BASE_REGS.R3, R4: BASE_REGS.R4 });
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_hex_keyboard_test.rs
use std::sync::Arc;

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

const KEY_BUF: u16 = 0x1800;
const MODE_FREE: u8 = 1;
const HEX_COLS: [u8; 8] = [0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80];

fn base_regs() -> CallRegisters {
    CallRegisters {
        r2: Some(0x2222),
        r3: Some(0x3333),
        r4: Some(0x4444),
        ..Default::default()
    }
}

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    s.run_init()?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

fn call_hex_key(
    s: &mut Mn1613AsmSession,
    mock: &Arc<IoBoardHandshakeMock>,
    buf_word_addr: u16,
) -> Result<(), FrameworkError> {
    mock.run_with_cpu_to_io_request(|| {
        let _ = s.call(
            "g_bios_hex_key_get_",
            CallOptions {
                registers: Some(CallRegisters {
                    r0: Some(buf_word_addr),
                    ..base_regs()
                }),
                ..Default::default()
            },
        )?;
        Ok(())
    })?;
    Ok(())
}

fn read_key_buf(s: &Mn1613AsmSession, word_addr: u16) -> [u16; 8] {
    [
        s.read_word(word_addr) & 0x00ff,
        s.read_word(word_addr + 1) & 0x00ff,
        s.read_word(word_addr + 2) & 0x00ff,
        s.read_word(word_addr + 3) & 0x00ff,
        s.read_word(word_addr + 4) & 0x00ff,
        s.read_word(word_addr + 5) & 0x00ff,
        s.read_word(word_addr + 6) & 0x00ff,
        s.read_word(word_addr + 7) & 0x00ff,
    ]
}

#[test]
fn free_mode_returns_hex_columns() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        {
            let mut st = mock.state.lock().expect("state lock");
            st.mode = MODE_FREE;
            st.hex_keys = HEX_COLS;
        }
        call_hex_key(s, &mock, KEY_BUF)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                ..Default::default()
            },
            None,
        )?;
        assert_eq!(
            read_key_buf(s, KEY_BUF),
            [0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80]
        );
        Ok(())
    })
}

#[test]
fn monitor_mode_returns_ng_and_zero_columns() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        {
            let mut st = mock.state.lock().expect("state lock");
            st.hex_keys = HEX_COLS;
        }
        call_hex_key(s, &mock, KEY_BUF)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(1),
                ..Default::default()
            },
            None,
        )?;
        assert_eq!(read_key_buf(s, KEY_BUF), [0, 0, 0, 0, 0, 0, 0, 0]);
        Ok(())
    })
}

#[test]
fn hex_key_get_preserves_r3_r4() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        {
            let mut st = mock.state.lock().expect("state lock");
            st.mode = MODE_FREE;
            st.hex_keys = HEX_COLS;
        }
        call_hex_key(s, &mock, KEY_BUF)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}
